src/app/services/favourites.service.ts

```diff
@@ -0,0 +1,106 @@
+import { Injectable, inject } from '@angular/core';
+import { Firestore, doc, getDoc, setDoc, arrayUnion, arrayRemove, updateDoc, DocumentReference, DocumentData } from '@angular/fire/firestore';
+import { User } from '@angular/fire/auth';
+import { AuthenticationService } from './authentication.service';
+import { Activity } from '../model/activity';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class FavouritesService {
+  private firestore: Firestore = inject(Firestore);
+  private currentUser: User | null;
+  private userDocRef: DocumentReference<DocumentData>;
+  private userFavoritesCache: string[];
+
+  constructor(private authService: AuthenticationService) {
+    this.setupAuthStatusListener();
+    this.loadFavouritesInCache();
+  }
+
+  private removeFromCache(activityId: string) {
+    const index = this.userFavoritesCache.indexOf(activityId);
+    if (index !== -1) {
+      this.userFavoritesCache.splice(index, 1);
+    }
+  }
+
+  private setupAuthStatusListener() {
+    // Subscribe to the authStatusSub BehaviorSubject in the AuthenticationService
+    this.authService.currentAuthStatus.subscribe(async (user) => {
+      this.currentUser = user;
+
+      if (this.currentUser) {
+        this.userDocRef = doc(this.firestore, 'users', this.currentUser.uid);
+      }
+    });
+  }
+
+  private async loadFavouritesInCache() {
+    const userDocSnapshot = getDoc(this.userDocRef);
+    this.userFavoritesCache = (await userDocSnapshot).get('favorites') || [];
+  }
+
+  async isActivityFavorited(activityId: string | null): Promise<boolean> {
+    if (!this.currentUser || activityId === null) {
+      throw new Error('User not logged in or activityId is null');
+    }
+
+    if (!this.userFavoritesCache || this.userFavoritesCache.length === 0) {
+      await this.loadFavouritesInCache();
```

Comment:
So if user doesn't have any activities favorited, it's going to keep loading for every activity. Can we remove this?

Reply:
Yeah this is not actually needed, have removed